[draft] String and Unicode processing functions

RedBearAK opened the pull request:
### Changes

Two new [draft] functions for processing simple alphanumeric strings (with spaces) and Unicode code point address references (between 1 and 6 hex characters). `ST()` for strings and `UC()` for Unicode addresses. Names will probably change. 

### Related issue thread

https://github.com/joshgoebel/keyszer/issues/43

joshgoebel commented:
Please have a look at tests and my simplifications:

- Use `Key.X` vs `C("X")`, this is simpler and avoids layers of complexity.
- I added digits to the `Key[]` lookup ability so they are no longer a weird special case
- Implemented unicode support for `type`.

I need some guidance on the Unicode entry stuff... can it indeed take an arbitrary # of hex digits (from 1 to 7)?  That's how it works now (see test)... there is no 0 padding, etc...

joshgoebel commented:
I'm not so sure that `unicode` shouldn't take a hex value rather than a string.... it's simpler and avoids the need for error checking since the compiler gives you that for free. IE:

```
unicode(0x1FC98)
```

Thoughts?

RedBearAK commented:
@joshgoebel 

> Please have a look at tests and my simplifications:

Trying to look through it. It's an extraordinary number of additions in a short period of time. What I can comprehend so far looks great. 

> Prefer Key.X vs C("X"), this is simpler and avoids layers of complexity.

Seems fine, but what about actual combos that need to be pressed together? This is only an example for a single key. Or, you mean the `C()` would still be used for combos, and this other form only for single keys? 

Love the numerics fix. Simplifies things. 

> I need some guidance on the Unicode entry stuff... can it indeed take an arbitrary # of hex digits (from 1 to 7)?

It's up to six digits, not seven. The test for length gt-eq 7 is for when to bail because it's too long. 

Apparently there are calculated gaps throughout the Unicode structure, with ways of checking that an address is legit. Exactly which of the shorter codes are actually real Unicode symbols, I'm not sure. But the manual method will at least show one of those square boxes with the code inside if even a single hex digit is used. So it doesn't seem to complain or fail to produce output. 

It may be that there are no legit single hex digit addresses. But I have seen a two-digit. 

This link seems to have a lot of relevant info: 

https://docs.microsoft.com/en-us/dotnet/standard/base-types/character-encoding-introduction

> I'm not so sure that unicode shouldn't take a hex value rather than a string.... it's simpler and avoids the need for [syntax] error checking since the compiler gives you that for free. IE:

> unicode(0x1FC98)

You're saying here you want to keep them separate? Or... oh, you want it to be explicitly a hex value instead of a string like `"00a1"`? Whatever is easiest and most reliable to process. The hex form is usually listed somewhere alongside the U+00A1 type formatting of the address. So it shouldn't be too much trouble or cause much confusion.

joshgoebel commented:
> Seems fine, but what about actual combos that need to be pressed together?

You can see I didn't change those... only simplified the individual `Key`s.

> oh, you want it to be explicitly a hex value

Well it could be a numeric period, someone writing it in hex would just be for there convenience if that was how they saw it in a reference, etc... (I updated my comment above)

RedBearAK commented:
> You can see I didn't change those... only simplified the individual Keys.

Makes sense. 

> Well it could be a numeric value, someone writing it in hex would just be for there convenience if that was how they saw it in a reference, etc... (I updated my comment above)

I see. So decimal or hex, just not as a string. I think that makes sense too. 

I noticed that a very generic function name like "string" conflicted with the `import string`. I worry that something as generic as `unicode()` may also run afoul of some other imported function. But you've got a lot more experience, so you'd know better if that could ever be a problem. 

So the tests, I'd just refresh the `venv` environment and then run the `test_api_helpers.py` in the terminal and it spits out the results of the tests to make sure everything is working? I haven't actually been in that environment for several days. The release version has been working quite nicely.

RedBearAK commented:
I guess I'd have to clone my own fork and checkout this branch in order to see & test this stuff before it's committed to your `main` or some testing branch of your own repo.

joshgoebel commented:
You already have you own fork if you made a PR.  You just need it local and then pull my latest changes.

RedBearAK commented:
Figured out how to checkout this PR into a new local branch: 

```
git fetch origin pull/60/head:unicode_strings
git checkout unicode_strings
```

So I can see the new `test_api_helpers.py`, and ran it in the `venv`. There's no output. I haven't actually used the test files before, so I'm not sure if I'm doing it right or what to expect. 

```
python3 ./test_api_helpers.py
```

RedBearAK commented:
This does some... stuff. How much of this needs to be fixed? 

```
(.venv)  kris@rbns-acer-3700u  ~/Downloads/josh-keyszer/keyszer   unicode_strings  ./tools/check.sh 
./tools/check.sh: line 4: codespell: command not found
./tools/check.sh: line 5: bandit: command not found
./tools/check.sh: line 6: flake8: command not found
./tools/check.sh: line 7: flake8: command not found
would reformat src/keyszer/cli.py
would reformat src/keyszer/config_api.py
would reformat src/keyszer/transform.py
Oh no! 💥 💔 💥
3 files would be reformatted, 18 files would be left unchanged.
./tools/check.sh: line 10: isort: command not found
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/kris/Downloads/josh-keyszer/keyszer, configfile: pytest.ini
plugins: looptime-0.2, asyncio-0.18.3
asyncio: mode=auto
collected 39 items                                                                                                                                                                                                

tests/test_api_helpers.py .........                                                                                                                                                                         [ 23%]
tests/test_basics.py ...........                                                                                                                                                                            [ 51%]
tests/test_helpers.py .                                                                                                                                                                                     [ 53%]
tests/test_keymap_basics.py ....                                                                                                                                                                            [ 64%]
tests/test_modmap.py ...                                                                                                                                                                                    [ 71%]
tests/test_modmap_cond.py ..s                                                                                                                                                                               [ 79%]
tests/test_modmap_cond_multi.py ..                                                                                                                                                                          [ 84%]
tests/test_modmap_multi.py ....                                                                                                                                                                             [ 94%]
tests/test_nested_keymaps.py .                                                                                                                                                                              [ 97%]
tests/test_sticky_bind.py .                                                                                                                                                                                 [100%]

========================================================================================== 38 passed, 1 skipped in 0.24s ==========================================================================================
./tools/check.sh: line 12: safety: command not found
```

joshgoebel commented:
Once you have all the requirements installed just run "pytest".

RedBearAK commented:
Output (without me changing anything but the command): 

```
(.venv)  kris@rbns-acer-3700u  ~/Downloads/josh-keyszer/keyszer   unicode_strings  pytest       
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/kris/Downloads/josh-keyszer/keyszer, configfile: pytest.ini
plugins: looptime-0.2, asyncio-0.18.3
asyncio: mode=auto
collected 39 items                                                                                                                                                                                                

tests/test_api_helpers.py .........                                                                                                                                                                         [ 23%]
tests/test_basics.py ...........                                                                                                                                                                            [ 51%]
tests/test_helpers.py .                                                                                                                                                                                     [ 53%]
tests/test_keymap_basics.py ....                                                                                                                                                                            [ 64%]
tests/test_modmap.py ...                                                                                                                                                                                    [ 71%]
tests/test_modmap_cond.py ..s                                                                                                                                                                               [ 79%]
tests/test_modmap_cond_multi.py ..                                                                                                                                                                          [ 84%]
tests/test_modmap_multi.py ....                                                                                                                                                                             [ 94%]
tests/test_nested_keymaps.py .                                                                                                                                                                              [ 97%]
tests/test_sticky_bind.py .                                                                                                                                                                                 [100%]

========================================================================================== 38 passed, 1 skipped in 0.22s ==========================================================================================
```

RedBearAK commented:
So `type` seems to be a pre-defined class (when I'm working with the `venv` VSCode loses the reference to the new stuff as it's importing from the current release version). You sure about going with a function name so generic? 

Obviously it's working when I actually run the patched branch, but it shows unrelated info about the object in VSCode. 

```
(class) type
type(object) -> the object's type type(name, bases, dict, **kwds) -> a new type
```

RedBearAK commented:
So this all works, both literal and escaped Unicode as part of otherwise simple upper/lower alphanumerics plus spaces. Only thing missing now is the rest of printable ASCII below 128. 

```
    C("RC-Alt-1"): unicode("1f339"), # U+1F339 = rose emoji
    C("RC-Alt-2"): type("\u00a1🌹Strings are useful 12345🌹\u00a1"),
    C("RC-Alt-3"): unicode("00a1"),
```

Nice not to need the `KEY_` anymore.

joshgoebel commented:
>  So type seems to be a pre-defined class 

It's a function, so yes we'll need a new name there.

joshgoebel commented:
Perhaps:

- `to_keystrokes("hello")
- `unicode_combo(0x1f323)`

joshgoebel commented:
> Only thing missing now is the rest of printable ASCII below 128.

Now regarding non-US keyboards... are there keyboards where `Shift-5` doesn't produce a %?  I mean we can't just say "type a %"... we have to say "shift-5"... so if that doesn't hold true universally I don't know how you'd convert ASCII non-alphnums into keycodes at all...

joshgoebel commented:
> I don't know how you'd convert ASCII non-alphnums into keycodes at all...

If that's true we could just always raise and explain that there isn't a canonical mapping across all keyboards for some characters, and they just wouldn't be supported by `to_keystrokes`

RedBearAK commented:
```
    to_keystrokes("hello")
    unicode_combo(0x1f323)
```

What about what I was originally calling the variable, too verbosely? `str_to_type` or maybe `str_to_keys`

Is the Unicode function intended to handle multiple Unicode keys at once? Or would `unicode_char` be OK? 

In its current form it still just wants the single reference at a time, max length 6. 

> Now regarding non-US keyboards... are there keyboards where Shift-5 doesn't produce a %?

That's the kind of thing I was really wondering about that AZERTY keyboard. On that keyboard it appeared that you get to the actual numbers with Shift, so `Shift-(` would be 5, instead of the other way around. And it's left parentheses instead of a percent sign. 

https://resource.logitech.com/content/dam/logitech/en/products/keyboards/mx-keys-for-business/gallery/mx-keys-business-keyboard-gallery-fra-graphite-1.png

Without having an actual international keyboard on hand, I don't know where to look for confirmation of this. But I feel like there must be existing tools that would allow the key reference set to adapt to the active keyboard layout for non-US keyboards, within reason.

joshgoebel commented on tests/test_api_helpers.py:
```diff
@@ -56,44 +56,59 @@ def test_combo_simple():
+
+    with pytest.raises(UnicodeNumberToolarge) as e:
+        out = unicode_keystrokes(0x110000)
+        assert e.message == "too large for Unicode keyboard entry."
```
This could probably be improved.

RedBearAK replied:
Not sure if this is related, or even an actual bug or something I did wrong in the config: 

```
Traceback (while executing your config):
  File "/home/kris/.config/kinto/keyszer.py", line 405, in <module>
  File "/home/kris/Downloads/josh-keyszer/keyszer/src/keyszer/config_api.py", line 181, in unicode_keystrokes
if (n > 0x10ffff):
TypeError: '>' not supported between instances of 'str' and 'int'
```

RedBearAK replied:
Of course I am. What kind of not-fool do you take me for? Remembered the function name but didn't change the strings. Fixed it.

joshgoebel commented:
> How much of this needs to be fixed?

None, you only need to worry about the checks reported to Github - which I'll likely turn up with time.

RedBearAK commented:
```

ASCII_TO_KEY = {
    ";": Key.SEMICOLON,
    "'": Key.APOSTROPHE,
    "=": Key.EQUAL,
    "-": Key.MINUS,
    "`": Key.GRAVE,
    "[": Key.LEFT_BRACE,
    "]": Key.RIGHT_BRACE,
    ",": Key.COMMA,
    ".": Key.DOT,
    "/": Key.SLASH,
    " ": Key.SPACE,
    "\\": Key.BACKSLASH
}
```
So the "uppercase" ASCII printables that you get to with Shift... the mapping is not so straightforward because the resulting character is... decided by the active keyboard layout? Am I thinking about this right? 

Is that why you aren't just putting the rest of the keyboard characters into this `dict` already?

RedBearAK commented:
Just messing around but I got the rest of the visible ASCII characters on my keyboard working with another `dict` in `config_api`, after making a few mistakes. Has to come below where `combo` is defined, apparently. But it works. 

Not submitting this as a change to the PR since I don't actually know which direction you're wanting to go to support the remaining ASCII characters. 

```python
# legacy helper name
K = combo
# short form for most common used helper
C = combo


def _create_modifiers_from_strings(modifier_strs):
    modifiers = []
    for modifier_str in modifier_strs:
        key = Modifier.from_alias(modifier_str)
        if key not in modifiers:
            modifiers.append(key)
    return modifiers


ASCII_WITH_SHIFT = {
    "~":    combo("Shift-Grave"),
    "!":    combo("Shift-1"),
    "@":    combo("Shift-2"),
    "#":    combo("Shift-3"),
    "$":    combo("Shift-4"),
    "%":    combo("Shift-5"),
    "^":    combo("Shift-6"),
    "&":    combo("Shift-7"),
    "*":    combo("Shift-8"),
    "(":    combo("Shift-9"),
    ")":    combo("Shift-0"),
    "_":    combo("Shift-Minus"),
    "+":    combo("Shift-Equal"),
    "{":    combo("Shift-Left_Brace"),
    "}":    combo("Shift-Right_Brace"),
    ":":    combo("Shift-Semicolon"),
    "\"":   combo("Shift-Apostrophe"),
    "<":    combo("Shift-Comma"),
    ">":    combo("Shift-Dot"),
    "?":    combo("Shift-Slash")
}


# ─── MARKS ──────────────────────────────────────────────────────────────────
```

Couple extra lines in the `for` loop in `to_keystrokes`: 

```python
    for c in s:
        if c.isupper():
            combo_list.append(combo("Shift-" + c))
        elif ord(c) > 127:
            combos = unicode_keystrokes(ord(c))
            combo_list.extend(combos)
        elif (str.isalnum(c)):
            combo_list.append(Key[c.upper()])
        elif c in ASCII_TO_KEY:
            combo_list.append(ASCII_TO_KEY[c])
        elif c in ASCII_WITH_SHIFT:
            combo_list.append(ASCII_WITH_SHIFT[c])
        else:
            raise CharacterNotSupported(f"The character {c} is not supported by `type` yet.")
```

joshgoebel commented:
> That's the kind of thing I was really wondering about that AZERTY keyboard. 

Do we/xkeysnail work on AZERTY keyboards or do we get the letters messed up?

RedBearAK commented:
The whole concept of whether the mapping is done in hardware or only in software via the keyboard layout setting is kind of breaking my brain. On the one hand, it would make a lot of sense to me if all keyboards since ancient times are actually "US" keyboards at the physical level, and are just remapped in software from the standard "US" key codes into the chosen non-US layout. But it may in fact work the other way. Meaning the software layout would have to know somehow that I'm on a physical AZERTY, in order to remap `A` to be a `Q` if I choose to use a US layout setting in software on the non-US hardware. Just because that seems to me like it would make keyboard layout files a lot more complicated doesn't mean that isn't exactly how it works. 

But while I ponder and google that, would it be possible to just let the user override the `ASCII_WITH_SHIFT` dictionary by placing a copy in the config file with different mappings? Or... maybe a `dict` with a different name, but the `to_keystrokes` function checks to see if that alternate `dict` is imported/populated and then uses that instead of the default mapping? If we could let the user override it would kind of become a moot point whether the default mapping from ASCII string character to Shift-key is universal or not. The user could just fix it for their layout. 

That could then evolve later on into a set of mappings for different layouts (hopefully submitted by users) being included in `config_api`, with the user only needing to pass in something like `LAYOUT = "azerty"` from the config to activate the right one. 

Of course that's if any of this is actually necessary.

joshgoebel commented:
I wrote a novel and then deleted it.  Too many unknowns.  I wish we have a few non-US people seriously interested in contributing...  

>  Meaning the software layout would have to know somehow that I'm on a physical AZERTY, in order to remap A to be a Q if I choose to use a US layout setting in software on the non-US hardware. 

Yes, this is a real issue... again I'm assuming we'd typically get the "REAL" keys keycodes... but if someone has a AZERTY keyboard they want to use as QWERTY, oh boy is that problematic...  I'm not sure any of this is fully solvable until we have a non-US champion who is more than trivially involved...  If the largest issue Kinto has had with such keyboards is the AltGr stuff that is promising though I think.

I really don't know how to understand some things on that AZERTY keyboard though like the `<>` key... on US keyboards that's two keys: the COMMA and DOT keys (by Linux naming standards)... and of course you need SHIFT to access the `<>` forms... so if [the keyboard] was trying it's HARDEST to be truthful with us I'm not sure what keycode it should send if one hit that keys all on it's own.  It seems there is no correct answer?

joshgoebel commented:
You should commit your `ASCII_WITH_SHIFT` lookup table but we should probably move both (and combine) into a single table in a new file, perhaps `ascii.py` in `lib`?  Our docs should note that `to_keycodes` may not work properly with non-US keyboards (and that issues are welcome).

RedBearAK commented:
> You should commit your ASCII_WITH_SHIFT lookup table but we should probably move both (and combine) into a single table in a new file, perhaps ascii.py in lib? Our docs should note that to_keycodes may not work properly with non-US keyboards (and that issues are welcome).

I was thinking about the same thing, especially if it will eventually need a bunch of different remappings, it would be cleaner in a separate file. I do all my commits through the GitHub web UI, so I'll have to leave it to you to move things into a new file. GitHub makes it very simple to edit an existing file and create a PR, but I'm still not sure how to create an entirely new file via the web UI. 

Not sure about combining into a single table. If there's going to be mapping issue, isn't it only going to be a problem with the combos and not the un-Shifted ASCII characters? Seems wise to leave the original `dict` separate for now. 

I looked around Amazon and Ebay for some sort of cheap AZERTY layout keyboard, and was somewhat shocked at how slim the pickings are. One keyboard in all of North America below $25, and it was a mid-2000s new-old-stock Apple USB keyboard (from back when they had the big clicky keys). 

But what does come up in spades in both places is sticker sets, for converting a regular keyboard to show the AZERTY layout. The fact these things are so common makes me lean toward the "all keyboards are US keyboards" theory again. 

Only way to know is to buy one.

joshgoebel commented:
> I do all my commits through the GitHub web UI,

Oh you really should get git locally, all sorts of nice UIs available.

RedBearAK commented:
@joshgoebel 

Trying to get a start on the Option-key scheme for Linux with the nested shortcuts. Things were going well until I tried to do some uppercase letters, I get a crash and traceback on startup. 

The `KeyError` appears to be referring to the accented character designated by the escaped Unicode string given to the `to_keystrokes` function, but this only seems to be a problem with Shift+key on the input side `C()`. 

```python
Traceback (while executing your config):
  File "/home/kris/.config/kinto/keyszer.py", line 395, in <module>
  File "/home/kris/Downloads/josh-keyszer/keyszer/src/keyszer/config_api.py", line 154, in to_keystrokes
combo_list.append(combo("Shift-" + c))
  File "/home/kris/Downloads/josh-keyszer/keyszer/src/keyszer/config_api.py", line 207, in combo
key = Key[key_str]
  File "/home/kris/Downloads/josh-keyszer/keyszer/src/keyszer/models/key.py", line 13, in __getitem__
return s.__getitem__(key)
  File "/usr/lib/python3.10/enum.py", line 437, in __getitem__
return cls._member_map_[name]
KeyError: 'È'
```

The related nested shortcut group. If I comment out the `Shift-e` line the crash doesn't happen. The lines not involving `Shift` work fine. 

```python
    C("Alt-Grave"): {
        C("Esc"):       ST("\u0060"),
        C("a"):         ST("\u00E0"),
        C("e"):         ST("\u00E8"),
        C("i"):         ST("\u00EC"),
        C("o"):         ST("\u00F2"),
        C("u"):         ST("\u00F9"),
        # C("Shift-a"):   ST("\u00C0"),
        C("Shift-e"):   ST("\u00C8"),
        # C("Shift-i"):   ST("\u00CC"),
        # C("Shift-o"):   ST("\u00D2"),
        # C("Shift-u"):   ST("\u00D9"),
    },
```

RedBearAK commented:
@joshgoebel 

I might have figured it out. Reversed the first two criteria in the `if-elif` tree, so that the "uppercase" Unicode character didn't escape by triggering `c.isupper()` before going through the Unicode function. 

```python
    combo_list = []
    for c in s:
        if ord(c) > 127:
            combos = unicode_keystrokes(ord(c))
            combo_list.extend(combos)
        elif c.isupper():
            combo_list.append(combo("Shift-" + c))
```

joshgoebel commented:
Commit the fix and push please.

RedBearAK commented:
> Commit the fix and push please.

Doesn't seem to cause any issues with my all-in-one test string, so... committed. 

552c1ef02005b6e0766e5439810d244f85c048a9

RedBearAK commented:
@joshgoebel 

Just accidentally discovered that the strings processor gets wonky if the Caps Lock key is active. At first I thought I messed something up with the recent commit, but everything was definitely working after I did that and I hadn't changed anything else. 

Same string output with Caps Lock OFF and ON: 

```
¡🌹Strings are useful ~!@#$%^&*()_+{}:"<>? ,.;'[]-=/\` 12345🌹¡

A1
1F339
sTRINGS ARE USEFUL ~!@#$%^&*()_+{}:"<>? ,.;'[]-=/\` 123451F339
A1
```

I suppose this was to be expected since the program is literally just "typing" at a low level, and this is what would happen if you were to try and type normally while the Caps Lock key is active. 

Seems like it will be slightly complicated avoiding this, since Shifting while Caps Lock is ON only reverses the affect on normal letter keys, while number keys and other keys pay no attention to Caps Lock.

The way it breaks Python's automatic detection of escaped (and literal) Unicode in the string is unfortunate. 

Oh no, this means my Option key special character scheme would break of the user has Caps Lock ON. 🤦🏽‍♂️ I was just using the main string function. I'll have to check and see if calling the Unicode function directly will also break with Caps Lock active.

RedBearAK commented:
@joshgoebel 

Maybe I'm overthinking it again. Is it possible to just detect Caps Lock ON, "press/release" Caps Lock to toggle it OFF, run the string/Unicode processor and then toggle Caps Lock back ON after the processor runs?

RedBearAK commented:
@joshgoebel 

Caps Lock state is affecting all output in general, of course. To me, this seems like an undesired outcome. 

```python
    C("Alt-Grave"): {                       # Dead Keys Accent Grave
        C("x"): C("z"),
```

This produces a lowercase "z" if Caps Lock is OFF, but an uppercase "Z" if Caps Lock is ON. Without invoking Shift+Z. 

With shortcuts this is never an issue (except for the Unicode shortcut) because normally the shortcut framework of the DE responds the same way regardless of the Caps Lock state. So this is only obvious when the output side of the keymapper is actually issuing a string rather than a shortcut combo. 

I'm trying to follow the online documentation for `evdev` and use the LED detection command (under "Getting and setting LED states"), but I keep getting "dev is not defined" even though I make sure to `import evdev`. This happens as part of a script or in the scratch environment, whether `keyszer` is running or not. I've tried many different things to try and get past this, but it always acts like `dev` just doesn't exist. 

This is supposed to provide a list of the available LEDs and their current state: 

```python
dev.leds(verbose=True)
```

joshgoebel commented:
`dev` isn't some magic variable... you need an **instance** of a device... you can find them in `input`... the name is passed thru into `transform` but we may need to start passing in the whole device... but you can at least add logging inside input to poke around at the devices.

I think we'd need to pass device into transform so it can be passed into `KeyState` so that KeyState would be able to query LEDs (to work with numlock)... and then if that works you could use numlock status as a conditional just like window name, etc...

RedBearAK commented:
> dev isn't some magic variable... you need an instance of a device

Ah. Thus why it is not defined. I see. How to determine when something in Python code is a function or variable or just some placeholder text is still giving me a lot of trouble. The tutorial provides no examples of using an actual device name to get or set the LEDs. 

Oh wait, there may be an example, but it's further down on the tutorial page, instead of above where they started using it. 

```python
dev = InputDevice('/dev/input/event1')
```

Maybe I can get somewhere with it now.

RedBearAK commented:
Well, no more `NameError` and I'm feeding it the correct device name for the keyboard, AFAIK, but the list always comes back empty with or without `verbose`. And with or without `keyszer` running. 

```python
Python 3.10.4 (main, Apr  2 2022, 09:04:19) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from evdev import *
>>> dev = InputDevice('/dev/input/event4')
>>> dev.leds(verbose=True)
[]
>>> dev.leds()
[]
>>> devices = [evdev.InputDevice(path) for path in evdev.list_devices()]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'evdev' is not defined
>>> import evdev
>>> devices = [evdev.InputDevice(path) for path in evdev.list_devices()]
>>> for device in devices:
...     print(device.path, device.name, device.phys)
... 
/dev/input/event15 Keyszer (virtual) Keyboard py-evdev-uinput
/dev/input/event14 HD-Audio Generic Front Headphone ALSA
/dev/input/event13 HD-Audio Generic HDMI/DP,pcm=3 ALSA
/dev/input/event12 HD User Facing: HD User Facing usb-0000:05:00.3-1/button
/dev/input/event8 ELAN0504:01 04F3:312A Touchpad i2c-ELAN0504:01
/dev/input/event7 ELAN0504:01 04F3:312A Mouse i2c-ELAN0504:01
/dev/input/event11 Acer WMI hotkeys wmi/input0
/dev/input/event10 Acer Wireless Radio Control acer-wireless/input0
/dev/input/event9 ETPS/2 Elantech Touchpad isa0060/serio1/input0
/dev/input/event6 Video Bus LNXVIDEO/video/input0
/dev/input/event5 Video Bus LNXVIDEO/video/input0
/dev/input/event4 AT Translated Set 2 keyboard isa0060/serio0/input0
/dev/input/event3 Power Button LNXPWRBN/button/input0
/dev/input/event2 Lid Switch PNP0C0D/button/input0
/dev/input/event1 Sleep Button PNP0C0E/button/input0
/dev/input/event0 Power Button PNP0C0C/button/input0
>>> dev.leds(verbose=True)
[]
>>> 
```

Must be some other fundamental thing I'm not understanding.

joshgoebel commented:
If your looping oever every device why not check all their LEDs inside that loop?

RedBearAK commented:
Like this? 

```python
>>> for device in devices:
...     device.leds(verbose=True)
...     print(device)
... 
[]
device /dev/input/event15, name "Keyszer (virtual) Keyboard", phys "py-evdev-uinput"
[]
device /dev/input/event14, name "HD-Audio Generic Front Headphone", phys "ALSA"
[]
device /dev/input/event13, name "HD-Audio Generic HDMI/DP,pcm=3", phys "ALSA"
[]
device /dev/input/event12, name "HD User Facing: HD User Facing", phys "usb-0000:05:00.3-1/button"
[]
device /dev/input/event8, name "ELAN0504:01 04F3:312A Touchpad", phys "i2c-ELAN0504:01"
[]
device /dev/input/event7, name "ELAN0504:01 04F3:312A Mouse", phys "i2c-ELAN0504:01"
[]
device /dev/input/event11, name "Acer WMI hotkeys", phys "wmi/input0"
[]
device /dev/input/event10, name "Acer Wireless Radio Control", phys "acer-wireless/input0"
[]
device /dev/input/event9, name "ETPS/2 Elantech Touchpad", phys "isa0060/serio1/input0"
[]
device /dev/input/event6, name "Video Bus", phys "LNXVIDEO/video/input0"
[]
device /dev/input/event5, name "Video Bus", phys "LNXVIDEO/video/input0"
[]
device /dev/input/event4, name "AT Translated Set 2 keyboard", phys "isa0060/serio0/input0"
[]
device /dev/input/event3, name "Power Button", phys "LNXPWRBN/button/input0"
[]
device /dev/input/event2, name "Lid Switch", phys "PNP0C0D/button/input0"
[]
device /dev/input/event1, name "Sleep Button", phys "PNP0C0E/button/input0"
[]
device /dev/input/event0, name "Power Button", phys "PNP0C0C/button/input0"
>>> 
```

joshgoebel commented:
But you only need to output the actual inputs devices you care about and keyszer (maybe), not all of them... and then fiddle with the locks WHILE running that in a loop, etc. see what happens

RedBearAK commented:
Oh, that's annoying. Bad assumption, as usual. I was sure it would show all of the available LED names and a zero if they were off. But I guess it's just going to be blank/null in the OFF state. I just did the same thing but with Caps Lock ON. 

```python
[('LED_CAPSL', 1)]
device /dev/input/event4, name "AT Translated Set 2 keyboard", phys "isa0060/serio0/input0"
```

But this is with both Caps Lock and Num Lock ON, slightly confusing: 

```python
[('LED_NUML', 0), ('LED_CAPSL', 1)]
device /dev/input/event4, name "AT Translated Set 2 keyboard", phys "isa0060/serio0/input0"
```

Setting `LED_NUML` to "0" is supposed to disable Num Lock. And it seems to. But the numbers in the output list appear to represent the key rather than the key state. As in, 0 = NumL, 1 = CapsL, 2 = ScrollL (guessing on the last one, I can't seem to activate Scroll Lock on my keyboard, and it's not shown in the tutorial). According to a ten year old post on StackOverflow from the `python-evdev` developer it seems like a good assumption. 

So I guess it's more of a matter of testing if LED_NUML (or zero) or LED_CAPSL (or one) appears in the list at all. 

The simpler output from `dev.leds()` without `verbose` is just the `ints`. 

```python
[0, 1]
```

So if `1` in `all(dev.leds())`, then Caps Lock is ON? And if `0` is there, Num Lock is ON. Something like that.

RedBearAK commented:
I verified this code at least does what I expect when I hard-code the list as `[0, 1]`, but when I try to use the `dev.leds()` from within the config file the list always comes out empty, even when both Lock keys are engaged when I restart `keyszer`. 

Maybe I need to stick it somewhere besides the config file, but seems like it should also work within the config. I'll try it in `config_api`. 

```python
import evdev
print("Trying to print LED states")
dev = InputDevice('/dev/input/event4')
LEDs_list = []
LEDs_list.extend(dev.leds())
print("LEDs_list is now:", LEDs_list)
if 1 in LEDs_list:
    print("Caps Lock is ON")
else:
    print("Caps Lock is OFF")
if 0 in LEDs_list:
    print("Num Lock is ON")
else:
    print("Num Lock is OFF")
print("After trying to print LED states")
```

joshgoebel commented:
>So I guess it's more of a matter of testing if LED_NUML (or zero) or LED_CAPSL (or one) appears in the list at all. ... So if 1 in all(dev.leds()), then Caps Lock is ON? And if 0 is there, Num Lock is ON. Something like that.

That was my understanding from looking in the past...

RedBearAK commented:
No luck moving the code into `config_api`. Still always has blank list output no matter the state of Caps Lock and Num Lock. I don't know what else to do with it for now. Still works when I go back into the python environment in the terminal. 🤷🏽‍♂️

joshgoebel commented:
```
>>> from evdev import InputDevice
>>> dev = InputDevice('/dev/input/event3')
>>> dev.leds()
[1]
```
Works fine for me if I open the device after the the caps lock state is ON.

RedBearAK commented:
Yep, works in the scratch environment or whatever you call it. In the terminal. And in the terminal inside VSCode. Just doesn't work from within the config or `config_api` files. 

I do the same import and setting of the variable either way. 

If I can't get it to even print out the correct state of the Lock keys because the list is always empty, I don't know how I could possibly do anything else with it.

joshgoebel commented:
> If I can't get it to even print out the correct state of the Lock keys because the list is always empty, I don't know how I could possibly do anything else with it.

I see your point, but it makes no sense.  There is no reason it shouldn't work inside keyszer, so my guess would be you're doing something wrong, something simple, but I have no idea what it is... so I'd work on adding the support DEEPER inside keyszer and see if it still doesn't work there... log the LEDs every time a key is pressed or something, etc... not just printing them once in the config.

RedBearAK commented:
Alright, I've seen it working, although the consistency of exactly when it works and when it comes out blank is pretty strange. Like the state is changing as it's getting read. I'm sure it's just because I'm not checking the state within the right context. 

But that means you're probably right about it working fine once it's in the right place. 

> log the LEDs every time a key is pressed or something, etc.

Yeah, I'll try something like that. Should be a good start.

joshgoebel commented:
>Yeah, I'll try something like that. Should be a good start.

And the easy way to do that is just go ahead and built it into KeyState in the first place... :-)

RedBearAK commented:
Are you referring to the `Keystate` class in `keystate.py`? 

I just tried putting some printing in `transform_key` and the output appears to be tracking the Lock keys states appropriately. Although I still have the issue of extremely stuttery log output. But they are switching from OFF to ON in the output. 

```python
dev = InputDevice('/dev/input/event4')


def transform_key(key, action, ctx):

    if 1 in dev.leds():
        print("(DD) ################  Caps Lock is ON", flush=True)
    else: 
        print("(DD) ################  Caps Lock is OFF", flush=True)
    if 0 in dev.leds():
        print("(DD) ################  Num Lock is ON", flush=True)
    else:
        print("(DD) ################  Num Lock is OFF", flush=True)
    
    global _active_keymaps
    is_top_level = False
```

```
(DD) ################  Caps Lock is OFF
(DD) ################  Num Lock is ON
(DD) on_key GRAVE release
(DD) on_key GRAVE press
(DD) ################  Caps Lock is OFF
(DD) ################  Num Lock is ON
(DD) on_key GRAVE release
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) suspending keys [RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.RIGHT_CTRL: 97>]
(DD) on_key TAB press
(DD) ################  Caps Lock is OFF
(DD) ################  Num Lock is ON
(DD) WM_CLS 'firefox' | DV 'AT Translated Set 2 keyboard' | KMAPS = [anonymous keymap, General Web Browsers, anonymous keymap, General GUI, Wordwise - not vscode]
(DD)   COMBO: RCtrl-TAB => [<ComboHint.BIND: 1>, Alt-TAB] [General GUI]
(DD) spent modifiers []
(DD) BIND: {<Key.RIGHT_CTRL: 97>: <Key.LEFT_ALT: 56>}
(DD) on_key TAB release
(DD) on_key TAB press
(DD) ################  Caps Lock is OFF
(DD) ################  Num Lock is ON
(DD) WM_CLS '' | DV 'AT Translated Set 2 keyboard' | KMAPS = [anonymous keymap, General GUI, Wordwise - not vscode]
(DD)   COMBO: RCtrl-TAB => [<ComboHint.BIND: 1>, Alt-TAB] [General GUI]
(DD) spent modifiers []
(DD) refusing to engage second sticky bind over existing sticky bind
(DD) on_key TAB release
(DD) on_key RIGHT_CTRL release
(DD) lift of BIND RIGHT_CTRL => LEFT_ALT
(DD) on_key CAPSLOCK press
(DD) ################  Caps Lock is OFF
(DD) ################  Num Lock is ON
(DD) on_key CAPSLOCK release
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) suspending keys [RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.RIGHT_CTRL: 97>]
(DD) on_key RIGHT_CTRL release
(DD) resume because of mod release
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) suspending keys [RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.RIGHT_CTRL: 97>]
(DD) on_key RIGHT_CTRL release
(DD) resume because of mod release
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) suspending keys [RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.RIGHT_CTRL: 97>]
(DD) on_key RIGHT_CTRL release
(DD) resume because of mod release


A1
A2
A3
?
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) suspending keys [RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.RIGHT_CTRL: 97>]
(DD) on_key RIGHT_CTRL release
(DD) resume because of mod release
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) suspending keys [RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.RIGHT_CTRL: 97>]
(DD) on_key RIGHT_CTRL release
(DD) resume because of mod release
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) suspending keys [RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.RIGHT_CTRL: 97>]
(DD) on_key RIGHT_CTRL release
(DD) resume because of mod release
(DD) on_key ENTER press
(DD) ################  Caps Lock is ON
(DD) ################  Num Lock is ON
(DD) on_key ENTER release
(DD) on_key ENTER press
(DD) ################  Caps Lock is ON
(DD) ################  Num Lock is ON
(DD) on_key ENTER release
(DD) on_key NUMLOCK press
(DD) ################  Caps Lock is ON
(DD) ################  Num Lock is ON
(DD) on_key NUMLOCK release
(DD) modmap: LEFT_META => LEFT_ALT [define_conditional_modmap (old API)]
(DD) on_key LEFT_ALT press
(DD) suspending keys [LAlt<Key.LEFT_ALT>]
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) resuspending keys
(DD) suspending keys [LAlt<Key.LEFT_ALT>, RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.LEFT_ALT: 56>, <Key.RIGHT_CTRL: 97>]
(DD) on_key LEFT_SHIFT press
(DD) on_key U press
(DD) ################  Caps Lock is ON
(DD) ################  Num Lock is OFF
(DD) WM_CLS 'Gnome-terminal' | DV 'AT Translated Set 2 keyboard' | KMAPS = [Special overrides for terminals, General GUI, Wordwise - not vscode, terminals]
(DD)   COMBO: LAlt-RCtrl-LShift-U => [Shift-SLASH, [Shift-Ctrl-U, <Key.A: 30>, <Key.KEY_1: 2>, <Key.ENTER: 28>], [Shift-Ctrl-U, <Key.A: 30>, <Key.KEY_2: 3>, <Key.ENTER: 28>], [Shift-Ctrl-U, <Key.A: 30>, <Key.KEY_3: 4>, <Key.ENTER: 28>], Shift-SLASH, ENTER] [General GUI]
(DD) spent modifiers []
(DD) on_key U release
(DD) on_key LEFT_SHIFT release
(DD) resume because of mod release
(DD) on_key RIGHT_CTRL release
(DD) resume because of mod release
(DD) on_key LEFT_ALT release
(DD) resume because of mod release
(DD) modmap: LEFT_ALT => RIGHT_CTRL [define_conditional_modmap (old API)]
(DD) on_key RIGHT_CTRL press
(DD) suspending keys [RCtrl<Key.RIGHT_CTRL>]
(DD) resuming keys: [<Key.RIGHT_CTRL: 97>]
```

joshgoebel commented:
Sorry maybe it's keycontext.  Whatever object we pass to the condition functions.

RedBearAK commented:
@joshgoebel 

> Now regarding non-US keyboards... are there keyboards where Shift-5 doesn't produce a %? I mean we can't just say "type a %"... we have to say "shift-5"... so if that doesn't hold true universally I don't know how you'd convert ASCII non-alphnums into keycodes at all...

So, as I discovered when experimenting with an Apple AZERTY (French/Belgian) keyboard, `evtest` will show the same key events on both US and non-US keyboard for the "same" key, and the key events don't change when changing the keyboard layout. This can be tested by just changing the layout setting on your own keyboard in your DE or Xorg settings. In other words, if the key would have been `RIGHT_BRACE` on a US keyboard, `evdev` will still see it as a right brace key event, even if the characters on the physical key on an AZERTY keyboard are Dollar/Asterisk/Euro (Normal/Shifted/Alt_Gr). 

So for non-Shifted, Shifted and Alt_Gr/Alt_Gr+Shift characters there would need to be separate ASCII mapping lists for each layout, or the string processor will output mismatched characters in many cases. 

Just as an example, the `< >` (left and right angled brackets) are on an extra key next to left Shift on the AZERTY that identifies as `KEY_102ND`. This is already in `key.py`. So the mapping for AZERTY would need to "type" Key_102nd and Shift+Key_102nd in order to output those characters from the string processor, rather than Shift+Comma and Shift+Dot as it would on a US layout. 

And the "5" key produces a left parentheses, Shift+5 will be the actual "5" character, and RAlt+5 will produce the left square brace (only if both Kinto and the DE are set up so that the Right Alt key acts as the special "Level3" character key). 

As noted in the Kinto thread on the AZERTY keyboard character issues, the changing of the layout doesn't seem to have any effect on how applications respond to shortcuts. It only seems to affect what characters appear in any kind of text entry area. 

Seems I was mistaken in thinking this might only be an issue with the Shifted ASCII characters, so it probably makes sense to combine the ASCII lists together. And it won't just be for symbols. Any letters that don't match the US layout (like the swapped A/Q characters on AZERTY) would also need to have their own mapping in the ASCII list for that layout, or they will produce the wrong character in output.

joshgoebel commented:
> ASCII_WITH_SHIFT 

Please see `test_api_helpers` and `test_ascii_keys`.... you should add a second test that tests ALL those new keystrokes you add to make sure they are translated properly.

I think after that i want to rename some of this stuff with a `us_` prefix to indicate it's US keyboard layout compatible - and then I think we could ship this in 0.7.

RedBearAK commented:
@joshgoebel 

There's another test that is failing, but the one I just added seems to be doing OK. 

```python
(.venv)  kris@rbns-acer-3700u  ~/Downloads/josh-keyszer/keyszer   unicode_strings ±  pytest
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/kris/Downloads/josh-keyszer/keyszer, configfile: pytest.ini
plugins: looptime-0.2, asyncio-0.18.3
asyncio: mode=auto
collected 43 items                                                                                                                                                                                                

tests/test_api_helpers.py .....F......                                                                                                                                                                      [ 27%]
tests/test_basics.py ............                                                                                                                                                                           [ 55%]
tests/test_helpers.py .                                                                                                                                                                                     [ 58%]
tests/test_keymap_basics.py ....                                                                                                                                                                            [ 67%]
tests/test_modmap.py ...                                                                                                                                                                                    [ 74%]
tests/test_modmap_cond.py ..s                                                                                                                                                                               [ 81%]
tests/test_modmap_cond_multi.py ..                                                                                                                                                                          [ 86%]
tests/test_modmap_multi.py ....                                                                                                                                                                             [ 95%]
tests/test_nested_keymaps.py .                                                                                                                                                                              [ 97%]
tests/test_sticky_bind.py .                                                                                                                                                                                 [100%]

==================================================================================================== FAILURES =====================================================================================================
_________________________________________________________________________________________ test_type_unsupported_character _________________________________________________________________________________________

    def test_type_unsupported_character():
        with pytest.raises(CharacterNotSupported) as e:
            out = to_keystrokes("{")
>           assert e.message == "The character { is not supported by `type` yet"
E           AttributeError: 'ExceptionInfo' object has no attribute 'message'

tests/test_api_helpers.py:69: AttributeError
============================================================================================= short test summary info =============================================================================================
FAILED tests/test_api_helpers.py::test_type_unsupported_character - AttributeError: 'ExceptionInfo' object has no attribute 'message'
===================================================================================== 1 failed, 41 passed, 1 skipped in 0.25s =====================================================================================
(.venv)  ✘ kris@rbns-acer-3700u  ~/Downloads/josh-keyszer/keyszer   unicode_strings ±  
```

This is what I put together. 

```python
def test_ascii_with_shift_keys():
    out = to_keystrokes('~!@#$%^&*()_+{}|:"<>?')
    assert [C("Shift-Grave"),C("Shift-1"),C("Shift-2"),C("Shift-3"),C("Shift-4"),
            C("Shift-5"),C("Shift-6"),C("Shift-7"),C("Shift-8"),C("Shift-9"),
            C("Shift-0"),C("Shift-Minus"),C("Shift-Equal"),C("Shift-Left_Brace"),
            C("Shift-Right_Brace"),C("Shift-Backslash"),C("Shift-Semicolon"),
            C("Shift-Apostrophe"),C("Shift-Comma"),C("Shift-Dot"),C("Shift-Slash")
             ] == out
```

And added the forgotten vertical bar to `ASCII_WITH_SHIFT`. 

```python
ASCII_WITH_SHIFT = {
    "~":    combo("Shift-Grave"),
    "!":    combo("Shift-1"),
    "@":    combo("Shift-2"),
    "#":    combo("Shift-3"),
    "$":    combo("Shift-4"),
    "%":    combo("Shift-5"),
    "^":    combo("Shift-6"),
    "&":    combo("Shift-7"),
    "*":    combo("Shift-8"),
    "(":    combo("Shift-9"),
    ")":    combo("Shift-0"),
    "_":    combo("Shift-Minus"),
    "+":    combo("Shift-Equal"),
    "{":    combo("Shift-Left_Brace"),
    "}":    combo("Shift-Right_Brace"),
    "|":    combo("Shift-Backslash"),
    ":":    combo("Shift-Semicolon"),
    "\"":   combo("Shift-Apostrophe"),
    "<":    combo("Shift-Comma"),
    ">":    combo("Shift-Dot"),
    "?":    combo("Shift-Slash")
}
```

If all this looks good I'll go ahead and submit it.

RedBearAK commented:
@joshgoebel 

> Well actually we should find ANOTHER unsupported character instead...

I suspected that test was no longer necessary. But I'm not sure what to use to replace it. All printable ASCII should be supported now. ASCII 00-31 and 127 (Delete) should all be blocked, but I guess we're not explicitly doing that yet. That would just leave the Extended ASCII characters. Everything above 127 is currently getting fed into `unicode_keystrokes`, but I'm not clear on whether the Extended ASCII characters actually work already, or only Unicode references. Haven't tried any ASCII that isn't actually on my keyboard. About half the characters from Extended ASCII are part of the Option key special characters on the Apple US layout. 

I'm wonder if `to_keystrokes` is really the right thing to label as US-specific. Isn't it just the specific lists like ASCII_TO_KEY and ASCII_WITH_SHIFT that should be re-labeled as layout-specific? Seems like `to_keystrokes` should just be a generic processor that uses info about the active (or defined in config file) layout to utilize the correct ASCII output lists for each layout. 

```python
ASCII_TO_KEYS_FR_AZERTY
ASCII_TO_KEYS_US
```

> There is not one-to-one correlation between languages, countries, and keyboard layouts.

I'm getting mixed results trying to look up a list of keyboard layout codes to use as examples. Seems the regional codes for layouts may not be completely standardized.

joshgoebel commented:
@RedBearAK Thanks for all the work on this!  I hope to release it soon in 0.7.
